pyro/infer/traceenum_elbo.py

```diff
@@ -20,6 +19,88 @@
+            if not (u < t):
+                continue
+            for v in log_factors:
+                if not (v < t):
+                    continue
+                if u <= v or v <= u:
+                    continue
+                left = ', '.join(sorted(f.name for f in u - v))
+                right = ', '.join(sorted(f.name for f in v - u))
+                raise ValueError("Expected tree-structured iarange nesting, but found "
+                                 "dependencies on independent iarange sets [{}] and [{}]"
+                                 .format(left, right))
+
+    # Check that all enumerated sites share a common subsampling scale.
+    # Note that we use a cheap weak comparison by id rather than tensor value, because
+    # (1) it is expensive to compare tensors by value, and (2) tensors must agree not
+    # only in value but at all derivatives.
+    if len(scales) != 1:
+        raise ValueError("Expected all enumerated sample sites to share a common poutine.scale, "
+                         "but found {} different scales.".format(len(scales)))
+
+
+def _contract(log_factors, enum_boundary):
+    # First close the set of ordinals under intersection (greatest lower bound),
+    # ensuring that the ordinals are arranged in a tree structure.
+    pending = list(log_factors)
+    while pending:
+        t = pending.pop()
+        for u in list(log_factors):
+            tu = t & u
+            if tu not in log_factors:
+                log_factors[tu] = []
+                pending.append(tu)
+
+    # Collect all enumeration dimensions.
+    enum_dims = defaultdict(set)
+    for t, terms in log_factors.items():
+        for term in terms:
+            for dim, size in enumerate(term.shape):
+                dim -= term.dim()
+                if dim >= enum_boundary:
+                    break
+                if size > 1:
+                    enum_dims[t].add(dim)
+
+    # Recursively combine terms in different iarange contexts.
+    while True:
+        leaf = max(log_factors, key=len)
+        terms = log_factors.pop(leaf)
+        dims = enum_dims.pop(leaf)
+
+        # Eliminate enumeration dims via a logsumproductexp() contraction.
+        shape = list(broadcast_shape(*set(x.shape for x in terms)))
+        remaining_dims = set.union(set(), *enum_dims.values())
+        for dim in dims - remaining_dims:
+            shape[dim] = 1
+        shape.reverse()
+        while shape and shape[-1] == 1:
+            shape.pop()
+        shape.reverse()
+        shape = tuple(shape)
+        term = logsumproductexp(terms, shape)
+        if not log_factors:
+            break
+
+        # Eliminate iarange dims via .sum() contractions.
+        frames = frozenset.intersection(*(leaf - t for t in log_factors if t < leaf))
+        assert frames
+        for frame in frames:
+            term = term.sum(frame.dim, keepdim=True)
+        parent = leaf - frozenset([frame])
+        log_factors[parent].append(term)
```

Comment:
Is there a possibility that `log_factors[parent]` does not exist. e.g. when we have nested iarange with no log factors in the `outer` iarange:

```python
with iarange("outer", ..):
    with iarange("inner", ..):
        pyro.sample(..)
```

Reply:
I may be missing something - I'm not sure if that will address this. Consider if `log_factors` consists of `frozenset()` and `frozenset(["a", "b"])`. That shouldn't change anything at the top. It seems like we are going to move up only one level in `parent = leaf - frozenset([frame])` and the remaining ordinal (either `frozenset(["a"])` or `frozenset(["b"])`) won't be in the list of `log_factors`.

Reply:
Oh you're right, that should be `parent = leaf - frames`. I'll add a test with a double jump that would have caught that.